sparse/coo/indexing.py

```diff
@@ -0,0 +1,444 @@
+    indices : tuple
+        The filtered indices.
+
+    Examples
+    --------
+    >>> _prune_indices((None, 5), (10,)) # None won't affect the mask
+    [5]
+    >>> _prune_indices((slice(0, 10, 1),), (10,)) # Full slices don't affect the mask
+    []
+    """
+    if prune_none:
+        indices = [idx for idx in indices if idx is not None]
+
+    i = 0
+    for idx, l in zip(indices[::-1], shape[::-1]):
+        if not isinstance(idx, slice):
+            break
+
+        if idx.start == 0 and idx.stop == l and idx.step == 1:
+            i += 1
+            continue
+
+        if idx.start == l - 1 and idx.stop == -1 and idx.step == -1:
+            i += 1
+            continue
+
+        break
+    if i != 0:
+        indices = indices[:-i]
+    return indices
+
+
+@numba.jit(nopython=True)
+def _compute_mask(coords, indices):  # pragma: no cover
+    """
+    Gets the mask for the coords given the indices in slice format.
+
+    Works with either start-stop ranges of matching indices into coords
+    called "pairs" (start-stop pairs) or filters the mask directly, based
+    on which is faster.
+
+    Exploits the structure in sorted coords, which is that for a constant
+    value of coords[i - 1], coords[i] is sorted. It uses this sortedness
+    to find sub-pairs for each dimension given the previous, and so on.
+    This is efficent for small slices, but not for large ones.
+
+    After it detects that working with pairs is rather inefficient (or after
+    going through each possible index), it constructs a filtered mask from the
+    start-stop pairs.
+
+    Parameters
+    ----------
+    coords : np.ndarray
+        The coordinates of the array.
+    indices : np.ndarray
+        The indices in the form of slices such that indices[:, 0] are starts,
+        indices[:, 1] are stops and indices[:, 2] are steps.
+
+    Returns
+    -------
+    mask : np.ndarray
+        The starts and stops in the mask.
+    is_slice : bool
+        Whether or not the array represents a continuous slice.
+
+    Examples
+    --------
+    Let's create some mock coords and indices
+
+    >>> import numpy as np
+    >>> coords = np.array([[0, 0, 1, 1, 2, 2]])
+    >>> indices = np.array([[0, 3, 2]])  # Equivalent to slice(0, 3, 2)
+
+    Now let's get the mask. Notice that the indices of ``0`` and ``2`` are matched.
+
+    >>> _compute_mask(coords, indices)
+    (array([0, 1, 4, 5]), False)
+
+    Now, let's try with a more "continuous" slice. Matches ``0`` and ``1``.
+
+    >>> indices = np.array([[0, 2, 1]])
+    >>> _compute_mask(coords, indices)
+    (array([0, 4]), True)
+
+    This is equivalent to mask being ``slice(0, 4, 1)``.
+    """
+    # Set the initial mask to be the entire range of coordinates.
+
+    starts = [0]
+    stops = [coords.shape[1]]
+    n_matches = coords.shape[1]
+
+    i = 0
+    while i < len(indices):
+        # Guesstimate whether working with pairs is more efficient or
+        # working with the mask directly
```

Comment:
I'm curious, what is the performance cost to just using one strategy or the other?  If it is not too great then we might consider simplifying to one.

Reply:
Unfortunately both are needed. In some cases, one is better, in some cases the other, and the difference is usually large.

For example, simple integer slicing mask construction is close to `O(log nnz)` in one and `O(nnz)` in the other.

And for slices such as `x[:500, :500, :500]`, the first approach boils down to `O(500^3)` worst case. However, if we work with pairs for the first two and switch to a mask later, it's a lot faster.

I made sure that the tests hit both cases, so we should be good. :-)